src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadStackFrame.cs

```diff
@@ -82,6 +82,15 @@ public JsonTypeInfo BaseJsonTypeInfo
         // Every required JsonPropertyInfo has RequiredPropertyIndex property which maps to an index in this BitArray.
         public BitArray? RequiredPropertiesSet;
 
+        // Represents known (non-extension) properties which have value assigned.
+        // Each bit corresponds to a property.
+        // False means that property is not set (not yet occurred in the payload).
+        // Length of the BitArray is equal to number of non-extension properties.
+        // Every JsonPropertyInfo has PropertyIndex property which maps to an index in this BitArray.
+        public BitArray? AssignedProperties;
```

Comment:
One optimization we left out when adding required property validation is avoiding allocations for a small number of properties. You could do that by implementing `VakyeBitArray` struct that either uses an int to track objects with less than 32 properties and fall back to allocating a regular `BitArray` for anything bigger. Not required for this PR but nice to have nonetheless.

Reply:
perhaps some simple solution like this could be good enough to significantly reduce allocations:

```
struct OptimizedBitArray
{
   private ulong _fastArray;
   private BitArray? _slowArray;

   public OptimizedBitArray(int numberOfProperties)
   {
     if (numberOfProperties > 64)
     {
         _slowArray = new BitArray();
     }
   }

   public void SetBit(int index)
   {
     if (_bitArray != null)
     // ...
   }

   // etc
}
```

Reply:
I'll add the data structure optimizations in separate PRs. I also want to add a `HasAllSet(BitArray mask)` overload which is useful for the required properties bitmask (might be good for a new API proposal too).

The other optimization I want to do is for `PropertyNameSet` to use a HashSet. The elements are `ReadOnlyMemory<byte>` in that case so we need to implement a custom comparer and it needs to use randomized Marvin seed to prevent collision attacks.